### PR TITLE
FoundationDB SQL Layer Adapter

### PR DESCRIPTION
**Pull request by gjvargas**
I've added support for a FoundationDB SQL Layer knex adapter. I tried my best to follow the style of code throughout the project. The FDB SQL layer code was adapted from the PostgreSQL code. I also updated the test directory to include both unit tests and integration tests for the fdbsql adapter.

For more information about the newly supported database, see here: https://foundationdb.com/

Pleas don't hesitate to contact me for questions and comments.

Thanks!

### CODE REVIEW

**Comment by bendrucker** [comment]
Please rebase this into a single commit when you have a chance and I'll review it. Knex's internals are starting to become unmanageable so please do the best you can to maximize test coverage. That's the only way to prevent regressions when this stuff needs to be rewritten.

**Comment by gjvargas** [comment]
I've rebased as requested. Nearly every test that previously existed for postgreSQL runs for the fdbsql adapter now as well. Note that the ones that aren't properly tested against the FoundationDB SQL Layer are testing unsupported operations for the database. A couple of examples include the tests using the JSON and enum types.

**Comment by tgriesser** [comment]
What Ben failed to mention though, is awesome work on putting this together @gjvargas!

Knex's internals are definitely getting a little hairy but should be looking really nice in the coming weeks as I'm going to dive back into things with a fresh look on modularity and coming up with an approach to use transducers throughout the codebase for consistent parsing and query generation.

All that said, Ben and I will take a look through and ask any questions.

File: lib/dialects/fdbsql/index.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+inherits(Client_FDB, Client);
+
+Client_FDB.prototype.dialect = 'fdbsql';
+
+// Lazy load the pg dependency, since we might just be using
+// the client to generate SQL strings.
+Client_FDB.prototype.initDriver = function() {
+  pg = pg || (function() {
+    try {
+      return require('pg');
```
Comment (tgriesser) [inline lib/dialects/fdbsql/index.js]: Is this supposed to be pg? or fdb?

Comment (gjvargas) [inline lib/dialects/fdbsql/index.js]: It's correct as is. FoundationDB actually uses the pg network protocol for convenience, so I was able to use the same dependency as the PostgreSQL adapter.

**Comment by gjvargas** [comment]
Awesome. Thanks guys!

**Comment by nathanlws** [comment]
Hi @tgriesser @bendrucker, I worked with @gjvargas on this and would be happy to make any needed changes. Anything I can do to help wrap this up?

We have ones ready to go for Bookshelf and Ghost once this is in :)

**Comment by bendrucker** [comment]
I'm fine with this. I'm restarting the last build to see what's up on the tests. As demand for other DBs increases, I'd like to generate multiple coverage reports specific to each DB so we can get a better handle on both DB-related status and stability.

**Comment by bendrucker** [comment]
Can you fix the tests?

**Comment by nathanlws** [comment]
> Can you fix the tests?

Sure!

SQL Layer isn't yet available in Travis so the integration tests will need to be skipped for CI. Is that ok?

**Comment by bendrucker** [comment]
Yeah, that goes along with the "experimental" comment.

**Comment by nathanlws** [comment]
Removed from default integration list and rebased. And looks like Travis is happy.

**Comment by bendrucker** [comment]
Great. On my list for the weekend.

**Comment by tgriesser** [comment]
Welp. Guess we'll see if there's any future here - https://www.npmjs.com/package/fdb

**Comment by bendrucker** [comment]
Wow — time to revert this then:

http://www.macrumors.com/2015/03/24/apple-acquires-foundationdb/

**Pull request merged**